### ISSUE TITLE
Refresh page causes logout

### ISSUE THREAD

**Issue by jmimi** [simplabs/ember-simple-auth#990]
I use my custom authenticator whch is the same as device authenticator almost.
When I refresh the page, local storage cleared and user logged out.

```
Ember version : 2.5.1
Ember Simple Auth version : 1.1.0
```

Why?
Thanks in advance

**Comment by jmimi** [comment]
`InternalSession.restore()` should be called once in initializing. But I didn't see anything.
I created a instance-initializer and calling that `restore` function, then everything is ok

**Comment by marcoow** [comment]
@jmimi: you should never call `InternalSession.restore` yourself - that's public API that might change any time.

**Comment by jmimi** [comment]
@marcoow I don't call `InternalSession.restore()` directly.
I have done this in my `session` service which inherited from ESA session service :
`this.get('session').restore();`

**Comment by marcoow** [comment]
Yep, that's what you should not do ;)

**Comment by jmimi** [comment]
@marcoow So what should I do to solve my problem?

**Comment by marcoow** [comment]
Can you post the code of your custom authenticator?

**Comment by jmimi** [comment]
It's so much similar to devise authenticator.

```
import Ember from 'ember';
import Base from 'ember-simple-auth/authenticators/base';
import config from 'my-app/config/environment';

const {RSVP: {Promise}, isEmpty, run, get, $} = Ember;

export default Base.extend({
    serverTokenEndpoint: Ember.computed(function () {
        return `${config.Backend.baseUrl}/${config.Backend.defaultNamespace}/accounts/login`;
    }),
    serverTokenEndpointLogout: Ember.computed(function () {
        return `${config.Backend.baseUrl}/${config.Backend.defaultNamespace}/accounts/logout`;
    }),

    restore(data) {
        const tokenAttribute = get(data, 'token');
        const identificationAttribute = get(data, 'username');

        if (!isEmpty(tokenAttribute) && !isEmpty(identificationAttribute)) {
            return Promise.resolve(data);
        } else {
            return Promise.reject();
        }
    },
    authenticate(identification, password) {
        return new Promise((resolve, reject) => {
            const data = {};
            data['password'] = password;
            data['username'] = identification;

            return this.makeRequest(data).then(
                (response) => run(null, resolve, response),
                (xhr) => run(null, reject, xhr.responseJSON || xhr.responseText)
            );
        });
    },
    invalidate() {
        return Promise.resolve();
    },

    makeRequest(data, options){
        const serverTokenEndpoint = this.get('serverTokenEndpoint');
        const requestOptions = $.extend({}, {
            url: serverTokenEndpoint,
            type: 'POST',
            dataType: 'json',
            data,
            beforeSend(xhr, settings) {
                xhr.setRequestHeader('Accept', settings.accepts.json);
                xhr.setRequestHeader('src', 'web');
            }
        }, options || {});

        return $.ajax(requestOptions);
    }
});

```

**Comment by marcoow** [comment]
Check that the `restore` method actually returns a resolving promise after reload - that's the reason for this kind of bug in 99% of the cases.

**Comment by jmimi** [comment]
When I reload the page, it seems `restore` isn't called

**Comment by marcoow** [comment]
Do you have an initializer or instance initializer in your app called `ember-simple-auth.js`? If so you need to remove that.

**Comment by jmimi** [comment]
No i don't.

**Comment by marcoow** [comment]
What's the content of the session store before the reload?

**Comment by jmimi** [comment]
```

authenticated: {username: 'blah', token: 'blahblah', type: 'blah'},
isAuthenticated: true

```
after reload, `autenticated` key gone and `isAuthentictaed` comes false

**Comment by marcoow** [comment]
Please copy/paste the actual store contents from `localStorage` or the cookie, depending on which one you are using.

**Comment by jmimi** [comment]
From `lcoalStorage` :
```
ember_simple_auth:session: {"locale":"en","authenticated":{"authenticator":"authenticator:myapp","username":"admin@myapp.xyz","token":"DrAdUHXSJRmqGxGmwtwr5BKDfalsrXA7fC1XJsD19EbsOFOs85NJrRAwOXYi9lJN","type":"password"}}
```

**Comment by marcoow** [comment]
Assuming you defined the above authenticator in `app/authenticators/myapp.js` everything looks good and the `restore` method should be called actually.

Would need access to the code to debug further. You can also place a breakpoint on [this line](https://github.com/simplabs/ember-simple-auth/blob/master/app/instance-initializers/ember-simple-auth.js#L6) and see why it never calls the authenticator's `restore` method.

**Comment by jmimi** [comment]
Oh, I have an instance initializer to set locale
```
import config from 'narengi/config/environment';

export function initialize(appInstance) {
    const i18n = appInstance.lookup('service:i18n');
    const session = appInstance.lookup('service:session');
    const locale = session.get('data.locale') || config.i18n.defaultLocale;
    i18n.set('locale', locale);
    if (!session.get('data.locale')) {
        session.set('data.locale', locale);
    }
}

export default {
    name: 'i18n',
    after: ['ember-simple-auth', 'ember-i18n'],
    initialize
};

```
This causes setting a key-value in session store and run before `setupSessionRestoration` then becuase `InternalSession` isn't setup so `authenticator` in `InternalSession` is null and causes storage to be cleared.

Now what should I do?
I set my instance initializer to be run after ESA. Where is my mistake?!

**Comment by marcoow** [comment]
Hm, this actually looks like a bug.

**Comment by jmimi** [comment]
Aha so I do my first solution for this temporarily.
Thanks a lot @marcoow

**Comment by marcoow** [comment]
No, you should never call `restore` manually.

Instead set the locale in the application route's `beforeModel` method. That only runs once the session has been restored successfully.

**Comment by jmimi** [comment]
Aha yes. You're right.
I'll do this.

**Comment by akhedrane** [comment]
any news about this bug?
how we can fix it temporary?

**Comment by marcoow** [comment]
@akhedrane: for now you cannot save anything in the session in an (instance) initializer. You can in the application route's `beforeModel` method though which should work for most of the use cases you'd want to use an initializer for (in many cases it's actually the better choice anyway).

**Comment by akhedrane** [comment]
i have initializer to inject i18n service only. no data saved in the session.
but the problem still exist.

**Comment by marcoow** [comment]
Doesn't sound like you're experiencing this bug then. You'll need to provide more info, code etc. so that your problem can be reproduced.

**Comment by swquinn** [comment]
I, too, am having issues where upon refreshing a page `ember-simple-auth` is losing the authentication context and I am seeing the `isAuthenticated` flag reset to false. I'm using the following

```
DEBUG: -------------------------------
DEBUG: Ember             : 2.7.1(unknown)
DEBUG: Ember Data        : 2.7.0(unknown)
DEBUG: jQuery            : 2.2.4(unknown)
DEBUG: Ember Simple Auth : 1.1.0(unknown)
DEBUG: -------------------------------
```

... along with [ember-simple-auth-token](https://github.com/jpadilla/ember-simple-auth-token) v1.1.1 for JWT support.

However, in my case, the only code that is inserting anything into the session is this addon, running `ack-grep` to search for instances where the session is being set I only found ember-simple-auth mixins and vendors, e.g.

> **Results of running ack-grep**
>
> ```bash
> vagrant@vagrant-ubuntu-trusty-64:/vagrant$ ack-grep --noignore=node_modules --js "set\([\'\"]session"
> 
> node_modules/ember-simple-auth/fastboot-dist/fastboot/vendor.js
> 87119:        this.set('session.attemptedTransition', null);
> 87209:          this.set('session.attemptedTransition', transition);
> 87829:        this.set('session.store.cookieExpirationTime', expirationTime);
> 
> node_modules/ember-simple-auth/fastboot-dist/assets/vendor.js
> 87119:        this.set('session.attemptedTransition', null);
> 87209:          this.set('session.attemptedTransition', transition);
> 87829:        this.set('session.store.cookieExpirationTime', expirationTime);
> 
> node_modules/ember-simple-auth/addon/mixins/authenticated-route-mixin.js
> 60:      this.set('session.attemptedTransition', transition);
> 
> node_modules/ember-simple-auth/addon/mixins/application-route-mixin.js
> 88:      this.set('session.attemptedTransition', null);
> 
> node_modules/ember-simple-auth/addon/session-stores/cookie.js
> 25:      this.set('session.store.cookieExpirationTime', expirationTime);
> 
> dist/assets/vendor.js
> 153675:        this.set('session.attemptedTransition', null);
> 153757:        this.set('session.attemptedTransition', transition);
> 154376:        this.set('session.store.cookieExpirationTime', expirationTime);
> ```

You have asked others in this thread what their session data looks like after a successful login, and what it looks like after refresh, so I will provide mine:

**After Login:**

After presenting a user's credentials and logging in, I am redirected to my index page per my configuration and the properties of the main session service, and the contents of the local storage are as follows:

```
// session:main
{
  isAuthenticated: true,
  authenticator: authenticator:jwt
}

// ember_simple_auth:session
{
  "authenticated": {
    "authenticator": "authenticator:jwt",
    "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJhcGkubG9yZWNhbGwuY29tIiwic3ViIjoiNTM2Y2M3YTk5MWE0Y2JhODE1MDAwMDRmIiwiaWF0IjoxNDcxNzEwOTU2LjMwNiwibmFtZSI6IlNlYW4gVy4gUXVpbm4iLCJ1c2VybmFtZSI6InN3cXVpbm4iLCJleHAiOjE0NzE3OTczNTZ9.OjTdfheZJgir_61am2dLoXZf_LIr3uZLWz755y_MkM8",
    "exp": 1471797356
  }
}
```

**After Refresh:**

After landing on the index page of my application I hit Command+R and the main session and local storage values change to:

```
// session:main
{
  authenticator: null,
  isAuthenticated: false
}

// ember_simple_auth:session
{
  "authenticated": {
    "authenticator": "authenticator:jwt",
    "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJhcGkubG9yZWNhbGwuY29tIiwic3ViIjoiNTM2Y2M3YTk5MWE0Y2JhODE1MDAwMDRmIiwiaWF0IjoxNDcxNzEwOTU2LjMwNiwibmFtZSI6IlNlYW4gVy4gUXVpbm4iLCJ1c2VybmFtZSI6InN3cXVpbm4iLCJleHAiOjE0NzE3OTczNTZ9.OjTdfheZJgir_61am2dLoXZf_LIr3uZLWz755y_MkM8",
    "exp": 1471797356
  }
}
```

If I can provide any more information that would be helpful, please let me know.

(As an aside, I seem to also be having an issue with my application's `beforeModel` hook seemingly not being called as well, but that's another issue for another day that I believe is unrelated to this add-on).

**Comment by anthonycollini** [comment]
I am having this issue as well.  I have found a pattern that causes it. A logged out user makes an unauthorized call the to the API, then logs in. The next refresh will log you out. I can't see anything in my code, or anything that jumps out that would cause this. My local storage looks fine before and after logging in. Be

Before logging in:
```
{"authenticated":{}}
```

After logging in:
```
{"authenticated":{"authenticator":"authenticator:oauth2-gateway","access_token":"50424dc00fe1fc526e5a537e16706b497c2ab4ea81542aab623eb9cefb264d08","token_type":"bearer","expires_in":600,"refresh_token":"383eeca9c191c034e73f353e90a9ea3ef1276591d48626456c7cf193dba3dde9","created_at":1477929759,"expires_at":1477930359010}}
```

**Comment by marcoow** [comment]
@anthonycollini: can you provide the source for your authenticator?

**Comment by anthonycollini** [comment]
I have 2 authenticators:

app/authenticators/oauth2-password-grant.js

```
import config from '../config/environment';
import OAuth2PasswordGrant from 'ember-simple-auth/authenticators/oauth2-password-grant';

export default OAuth2PasswordGrant.extend({

  clientId: config.oauth2ClientId,
  serverTokenEndpoint: config.apiURL + '/oauth/token',
  serverTokenRevocationEndpoint: config.apiURL + '/oauth/revoke',

});
```

app/authenticators/oauth2-gateway.js
```
import config from '../config/environment';
import Ember from 'ember';
import OAuth2PasswordGrant from 'ember-simple-auth/authenticators/oauth2-password-grant';

export default OAuth2PasswordGrant.extend({

  torii: Ember.inject.service(),
  clientId: config.oauth2ClientId,
  serverTokenEndpoint: config.apiURL + '/oauth/token',
  serverTokenRevocationEndpoint: config.apiURL + '/oauth/revoke',

  authenticate(torii_provider) {
    return new Ember.RSVP.Promise((resolve, reject) => {
      this.get('torii').open(torii_provider).then(data => {
        return this.makeRequest(config.apiURL + '/oauth/gateway_authorize', data);
      }).then(response => {
        // Copied from OAuth2PasswordGrant#authenticate
        Ember.run(() => {
          const expiresAt = this._absolutizeExpirationTime(response['expires_in']);
          this._scheduleAccessTokenRefresh(response['expires_in'], expiresAt, response['refresh_token']);
          if (!Ember.isEmpty(expiresAt)) {
            response = Ember.assign(response, { 'expires_at': expiresAt });
          }
          resolve(response);
        });
      }).catch(error => {
        reject(error);
      });
    });
  }

});
```

**Comment by marcoow** [comment]
@anthonycollini: do you have any custom initializers that access the session as described by @jmimi above? If not please create a new issue for your problem as it's most likely unrelated then.

**Comment by jcbvm** [comment]
@marcoow is this 'bug' only present when setting a value on the session or accessing it before the beforeModel hook has fired? I'm having the same problem, but for me it's hard te reproduce, seems like a timing issue (it was working before for a long time, same version 1.1.0).

**Comment by marcoow** [comment]
@jcbvm: would be good if you could share some of the code so we can have a closer look at what might be happening.

**Comment by jcbvm** [comment]
@marcoow below some code, stripped some sensitive data

```javascript 
// custom authenticator 'main'

import Ember from 'ember';
import Base from 'ember-simple-auth/authenticators/base';

export default Base.extend({
    store: Ember.inject.service(),
    userService: Ember.inject.service(),

    authenticate(credentials) {
        return this.get('userService').authenticate(credentials.username, credentials.password).then(data => {
            return this.didAuthenticate(data);
        });
    },
    restore() {
        return this.get('userService').getAuthentication().then(data => {
            return this.didAuthenticate(data);
        });
    },
    invalidate() {
        return this.get('userService').clearAuthentication();
    },
    didAuthenticate(data) {
        // stripped some code here, it basicly pushes some data in store and returns session data
        return Ember.RSVP.resolve();
    }
});
```
```javascript
// login route

actions: {
        authenticate() {
            let credentials = this.controller.getProperties('username','password');
            this.session.authenticate('authenticator:main', credentials);
        }
}
```
```javascript
// application route

import Ember from 'ember';
import ApplicationRouteMixin from 'ember-simple-auth/mixins/application-route-mixin';

export default Ember.Route.extend(ApplicationRouteMixin, {
    session: Ember.inject.service(),

    beforeModel(transition) {
        return Ember.RSVP.cast(this._super(...arguments)).then(() => {
            if (this.get('session.isAuthenticated')) {
                this.sessionRestored();
            }
            let lang = this.get('session.data.lang');
            // set the language based on session lang
        });
    },
    sessionRestored() {
        // do some stuff unrelated to session
        let appController = this.controllerFor('application');
        // some property from appController is read here, if `false` is returned, the session will be invalidated.
        // I first thought the issue was here if appController is still undefined, 
        // but from what I can see this is not the case, otherwise I would see an error appear.
    },
    ...
});
```
```javascript
// session-service initializer

export default {
    initialize(application) {
        application.inject('model', 'session', 'service:session');
        application.inject('route', 'session', 'service:session');
        application.inject('controller', 'session', 'service:session');
        application.inject('component', 'session', 'service:session');
        application.inject('template', 'session', 'service:session');
    }
};
```

**Comment by jcbvm** [comment]
@marcoow Just a little note, I'm encountering this a lot of times with auto refresh when developing, same behaviour, restore gets never been called.

**Comment by marcoow** [comment]
@jcbvm: what's the code of the `userService`? Which session store are you using?

**Comment by jcbvm** [comment]
@marcoow userService just makes an ajax call to my backend, nothing special there. I'm using the localStorage Session Store.

**Comment by marcoow** [comment]
@jcbvm: I don't see anything in your code that's wrong. Would be helpful if you could setup a demo app that illustrates the problem you have in your app.

**Comment by jcbvm** [comment]
@marcoow I'm currently trying to narrow down the issue. It seems like I only have this issue when livereloading is on in development mode. Not been able to reproduce this by F5-ing via the browser.

**Comment by kevlarr** [comment]
I'm running into this (or a very similar..?) issue, only I can replicate it by manually refreshing the browser - `restore` gets called only the first time I refresh and not on subsequent page reloads.

```
import Ember from 'ember';
import Base from 'ember-simple-auth/authenticators/base';

const { RSVP, inject: { service }, setProperties } = Ember;

export default Base.extend({
  session: service('session'),
  ajax: service(),

  authenticate(email, password) {
    return new RSVP.Promise((resolve, reject) => {
      this.get('ajax').post('/api/login', {
        data: { email, password }
      }).then(response => {
        response.userId = response.user_id;
        delete response.user_id;

        const { userId, groups, preferences } = response;
        setProperties(this.get('session.data'), { userId, groups, preferences });

        resolve(response);
      }).catch(error => reject(error));
    });
  },

  invalidate() {
    return new RSVP.Promise((resolve, reject) => {
      this.get('ajax').post('/api/logout').then(() => {
        const data = this.get('session.data');
        ['userId', 'groups', 'preferences'].forEach(prop => delete data[prop]);

        resolve();
      }).catch(error => reject(error));
    });
  },

  restore({ userId, groups, preferences } = {}) {
    // Must have server cookie present in addition to session data
    const hasCookie = document.cookie.indexOf('cookie_name') >= 0;

    if (hasCookie && userId && groups && preferences) {
      return RSVP.Promise.resolve(arguments);
    }

    return RSVP.Promise.reject();
  },
});
```

I don't have any initializers (instance or otherwise), and I don't even have a `routes/application` - I'm just relying on the one generated by `ember-simple-auth`

**Comment by mansona** [comment]
Ok so we have just hit this issue and I have tracked down the problem, our code was actually throwing an Exception in our authenticator's `restore` method (details of which are not too important) 

The problem is that the error is being swallowed and not ever output to the user. The problem is that if there is an error thrown at this point https://github.com/simplabs/ember-simple-auth/blob/master/addon/internal-session.js#L67 it will not be caught by this error handler: https://github.com/simplabs/ember-simple-auth/blob/master/addon/internal-session.js#L85 , instead that error handler will only catch issues with the `this._callStoreAsync('restore')` call on this line https://github.com/simplabs/ember-simple-auth/blob/master/addon/internal-session.js#L62

If the error handler is valid for cases of exceptions thrown in the **user provided restore method** then I would recommend changing the `internal-session` `restore` method to look more like this: 

```
restore() {
    this._busy = true;
    const reject = () => RSVP.Promise.reject();

    return this._callStoreAsync('restore').then((restoredContent) => {
      let { authenticator: authenticatorFactory } = restoredContent.authenticated || {};
      if (authenticatorFactory) {
        delete restoredContent.authenticated.authenticator;
        const authenticator = this._lookupAuthenticator(authenticatorFactory);
        return authenticator.restore(restoredContent.authenticated).then((content) => {
          this.set('content', restoredContent);
          this._busy = false;
          return this._setup(authenticatorFactory, content);
        }, (err) => {
          debug(`The authenticator "${authenticatorFactory}" rejected to restore the session - invalidating…`);
          if (err) {
            debug(err);
          }
          this._busy = false;
          return this._clearWithContent(restoredContent).then(reject, reject);
        });
      } else {
        delete (restoredContent || {}).authenticated;
        this._busy = false;
        return this._clearWithContent(restoredContent).then(reject, reject);
      }
    }).then(null, () => {
      this._busy = false;
      return this._clear().then(reject, reject);
    });
  },
```
The difference is that this will now catch all issues with **user provided restore methods**

**Comment by mansona** [comment]
@jcbvm related to my comment above, it could be that you are getting an error along the lines of `Cannot call method 'authenticate' of undefined` and the error is being swallowed.

From the code you provided I wonder have you got a service called 'user'? if so your inject might need to look more like this: 
```
userService: Ember.inject.service('user'),
```
but that's just a stab in the dark 😂

**Comment by lougreenwood** [comment]
I'm hitting a similar problem, albeit on the third attempt at reloading the app when authenticated:

* Login
* Reload 1
* `Authenticator.restore()` called, user is authenticated
* Reload 2
* `Authenticator.restore()` called, user is authenticated
* Reload 3
* `Authenticator.restore()` **NOT** called, user is kicked out of the app...

I don't have any initialisers, but I do override the session service. In the override I do persist some session state in `session.init()` - not sure if this is similar to using an initialiser...

**Comment by lougreenwood** [comment]
Also, it seems that if I move the code for setting the values in the store from the `session.init()` method to `beforeModelI()` in the application route, the first refresh causes the session to invalidate

**Comment by lougreenwood** [comment]
Further update... after successfully logging in, if I temporarily disable the code in `session.init()` which writes to the store, then I don't trigger this bug...

Here's what my  `session.init()` looks like:

```
export default SessionService.extend({
    async init() {
        this._super(...arguments)

        // !! Comment out this line to not trigger refresh logout bug
        await this._setDefaults()
    },

    /**
     * Sets default `state` & `code_verifier`.
     * Intended to be used on session service init so that all app components
     * which generate authorize links use the same `code_verifier`
     */
    async _setDefaults() {
        // Ensure all invocations use the same state & code_verifier
        let { code_verifier } = await this.session.store.restore()
        const state         = this._generateState()

        if(!code_verifier) {
            code_verifier = this._generateCodeVerifier()
        }

        await this.session.store.persist({ state, code_verifier })
    },
```

**Comment by marcoow** [comment]
@lougreenwood: you cannot write to the session store before the session has been fully created.

**Comment by lougreenwood** [comment]
Sure, but shouldn't the session store have been created during the the session service is init'd - that appears to be a reasonable expectation?

In any case, I worked around this as much as I need to by using custom properties to hold default values. Seems this addresses my problem and maybe it'll help other people in future who come across a similar issue...

**Comment by marcoow** [comment]
The session subscribes to update events on the session stores and will try to restore itself whenever the store changes. That will very likely lead to problem if that happens at the wrong time. In general, you should never access the store directly.

**Comment by lougreenwood** [comment]
> In general, you should never access the store directly.

Not sure I follow, do you mean that the following interaction with the store is to be avoided?
* `let { code_verifier } = await this.session.store.restore()`
* `await this.session.store.persist({ state, code_verifier })`

**Comment by marcoow** [comment]
yes, exactly

**Comment by lougreenwood** [comment]
ok - I don't mean to be terse, but if it's not supposed to be used (or if the safe use cases for it are vague/minimal), why is it public?

Also, if you don't mind, could you describe how to safely store session state pls?

Seems that this long-living bug is likely a mis-understanding of how ESA should be used and the purpose of various public APIs?

**Comment by marcoow** [comment]
You can store custom session data by setting it on the session:

```js
this.session.set('data.authenticated.myData', 'maVal');
```

That is how it currently works but also the reason for the problems I mentioned above - basically all data that is written to the session is **automatically** persisted to the store which leads to a lot of hacky and complex code. This needs to be refactored into a cleaner solution but as that's a larger effort nobody has yet found the time to do it.

**Comment by marcoow** [comment]
cleaning up old issues
